test/xkeyboard-config-test.py.in

```diff
@@ -79,36 +86,60 @@ def xkbcomp(r='evdev', m='pc105', l='us', v='', o=''):
         setxkbmap.stdout.close()
         stdout, stderr = xkbcomp.communicate()
         if xkbcomp.returncode != 0:
-            print('ERROR: Failed to compile: {}'.format(' '.join(args)))
+            outstr.append('ERROR: Failed to compile: {}'.format(' '.join(args)))
+            success = False
         if xkbcomp.returncode != 0 or verbose:
-            print(stdout.decode('utf-8'))
-            print(stderr.decode('utf-8'))
+            outstr.append(stdout.decode('utf-8'))
+            outstr.append(stderr.decode('utf-8'))
 
     # This catches setxkbmap errors.
     except subprocess.CalledProcessError as err:
-        print('ERROR: Failed to compile: {}'.format(' '.join(args)))
-        print(err.output.decode('utf-8'))
+        outstr.append('ERROR: Failed to compile: {}'.format(' '.join(args)))
+        outstr.append(err.output.decode('utf-8'))
+        success = False
+
+    return success, '\n'.join(outstr) if outstr else None
 
 
-def parse(root, tool):
+def parse(path):
+    root = ET.fromstring(open(path).read())
     layouts = root.findall('layoutList/layout')
 
     options = [
         e.text
         for e in root.findall('optionList/group/option/configItem/name')
     ]
 
-    for l in progress_bar(layouts, 'layout '):
+    combos = []
+    for l in layouts:
         layout = l.find('configItem/name').text
-        tool(l=layout)
+        combos.append({'l': layout})
 
         variants = l.findall('variantList/variant')
-        for v in progress_bar(variants, 'variant'):
+        for v in variants:
             variant = v.find('configItem/name').text
-            tool(l=layout, v=variant)
 
-            for option in progress_bar(options, 'option '):
-                tool(l=layout, v=variant, o=option)
+            combos.append({'l': layout, 'v': variant})
+            for option in options:
+                combos.append({'l': layout, 'v': variant, 'o': option})
+
+    return combos
+
+
+def run(combos, tool, njobs):
+    failed = False
+    with Pool(njobs) as p:
+        results = []
+        for combo in progress_bar(combos, 'preparing'):
```

Comment:
Using `p.imap()` will save the 'preparing' phase and will be a bit more memory efficient.

Reply:
fwiw, the reason I didn't go with `imap` initially was that it doesn't take the `kwargs` as nicely. But then I realised it's just a test suite anyway so we can change the tool functions to take the dict we pass through. so ... done